### ISSUE TITLE
Failed to compile Serializer companion on Native

### ISSUE THREAD

**Issue by thevery** [Kotlin/kotlinx.serialization#290]
JVM (Android) ok
Serialization 0.9.1 / kotlin 1.3.10

Exception:
```
> Task :compileKotlinIos
error: compilation failed: org.jetbrains.kotlin.ir.declarations.lazy.IrLazyFunction@76a6f045

 * Source files: Sample.kt, SampleIos.kt
 * Compiler version info: Konan: 1.0.2 / Kotlin: 1.3.0
 * Output kind: LIBRARY

exception: java.lang.IllegalStateException: org.jetbrains.kotlin.ir.declarations.lazy.IrLazyFunction@76a6f045
        at org.jetbrains.kotlin.backend.konan.llvm.LlvmDeclarations.forFunction(LlvmDeclarations.kt:44)
        at org.jetbrains.kotlin.backend.konan.llvm.ContextUtils$DefaultImpls.getLlvmFunction(ContextUtils.kt:164)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGenerator.getLlvmFunction(CodeGenerator.kt:26)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGenerator.llvmFunction(CodeGenerator.kt:28)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.callDirect(IrToBitcode.kt:2536)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateSimpleFunctionCall(IrToBitcode.kt:2152)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateFunctionCall(IrToBitcode.kt:2104)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateCall(IrToBitcode.kt:1845)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateExpression(IrToBitcode.kt:822)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateExplicitArgs(IrToBitcode.kt:1986)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateCall(IrToBitcode.kt:1837)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.evaluateExpression(IrToBitcode.kt:822)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.generateVariable(IrToBitcode.kt:1272)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.generateStatement(IrToBitcode.kt:859)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitFunction(IrToBitcode.kt:736)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitSimpleFunction(IrElementVisitorVoid.kt:52)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitSimpleFunction(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitSimpleFunction(IrElementVisitorVoid.kt:53)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitSimpleFunction(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitSimpleFunction(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFunctionImpl.accept(IrFunctionImpl.kt:87)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:247)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:776)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.declarations.impl.IrClassImpl.accept(IrClassImpl.kt:102)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:247)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:776)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitClass(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.declarations.impl.IrClassImpl.accept(IrClassImpl.kt:102)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.acceptChildren(IrFileImpl.kt:79)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:251)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitFile(IrToBitcode.kt:522)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:38)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitFile(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitFile(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.accept(IrFileImpl.kt:76)
        at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.acceptChildren(IrModuleFragmentImpl.kt:45)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:251)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:379)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:28)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:291)
        at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.accept(IrModuleFragmentImpl.kt:42)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:247)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$7.invoke(IrToBitcode.kt:162)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$7.invoke(IrToBitcode.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:137)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:111)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:37)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase(KonanPhases.kt:136)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt.emitLLVM(IrToBitcode.kt:161)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$6$2.invoke(KonanDriver.kt:95)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$6$2.invoke(KonanDriver.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:137)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:111)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:37)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase(KonanPhases.kt:136)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$6.invoke(KonanDriver.kt:94)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$6.invoke(KonanDriver.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:137)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:111)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:37)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase(KonanPhases.kt:136)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:88)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:79)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:35)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:96)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:52)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:93)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:71)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:39)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:202)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:194)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:214)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:205)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:37)
        at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:31)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:207)
        at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:259)
        at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:16)
```

Minimal sample:
```
@Serializable
data class NotePatch(
    val title: String? = null
) {
    @Serializer(forClass = NotePatch::class)
    companion object : KSerializer<NotePatch> {
        override fun serialize(output: Encoder, obj: NotePatch) {
            val elemOutput = output.beginStructure(descriptor)
            elemOutput.endStructure(descriptor)
        }
    }
}
```

**Comment by sandwwraith** [comment]
Workaround: manually implement `deserialize` and `descriptor` from `KSerializer`.

**Comment by sandwwraith** [comment]
Fixed in 1.3.20
